src/persist-client/src/internal/gc.rs

```diff
@@ -205,273 +209,224 @@ where
-            // then we should be able to delete TypedStateVersionsIter (and
-            // probably merge UntypedStateVersionsIter into StateVersionsIter).
-            .check_codecs::<K, V, D>()
-            .expect("shard codecs should not change");
+            .expect("state is initialized")
+            .check_ts_codec()
+            .expect("ts codec has not changed");
+        report_step_timing(&machine.applier.metrics.gc.steps.fetch_seconds);
+
+        machine
+            .applier
+            .shard_metrics
+            .gc_live_diffs
+            .set(u64::cast_from(states.len()));
 
         debug!(
-            "gc {} for {} got {} versions from scan",
-            req.shard_id,
+            "gc seqno_since: ({}) got {} versions from scan",
             req.new_seqno_since,
             states.len()
         );
-        report_step_timing(&machine.applier.metrics.gc.steps.fetch_seconds);
 
-        let earliest_live_seqno = match states.peek_seqno() {
-            Some(x) => x,
-            None => panic!(
-                "initialized shard {} should have at least one live state",
-                req.shard_id
-            ),
-        };
-        // Fast-path: Someone already GC'd past `req.new_seqno_since`, don't
-        // bother running any of the below logic.
-        //
-        // Also a fix for #14580.
-        if earliest_live_seqno > req.new_seqno_since {
-            machine.applier.metrics.gc.noop.inc();
-            debug!(
-                "gc {} early returning, already GC'd past {}",
-                req.shard_id, req.new_seqno_since,
+        Self::incrementally_gc(
+            &req.shard_id,
+            &mut states,
+            req.new_seqno_since,
+            machine,
+            &mut report_step_timing,
+        )
+        .await;
+
+        // WIP: do we want to remove rollups incrementally?
+        let removable_rollups = machine.applier.removable_rollups();
+        let (removed_rollups, maintenance) = machine.remove_rollups(&removable_rollups).await;
+        debug!("removed rollups: {:?}", removed_rollups);
+
+        // WIP: do we want to calculate the seqno_parts_held metric?
+        while let Some(_) = states.next(|_| {}) {}
+
+        maintenance
+    }
+
+    /// Physically deletes all blobs from Blob and live diffs from Consensus that
+    /// are safe to delete, given the `seqno_since`, ensuring that the earliest
+    /// live diff in Consensus has a rollup of seqno `<= seqno_since`.
+    ///
+    /// Internally, performs deletions for each rollup encountered, ensuring that
+    /// incremental progress is made even if the process is interrupted before
+    /// completing all gc work.
+    async fn incrementally_gc<F>(
+        shard_id: &ShardId,
+        states: &mut StateVersionsIter<T>,
+        seqno_since: SeqNo,
+        machine: &mut Machine<K, V, T, D>,
+        timer: &mut F,
+    ) where
+        F: FnMut(&Counter),
+    {
+        let delete_semaphore = Semaphore::new(
```

Comment:
any advantage to having the semaphore here instead of constructing it in `delete_and_truncate` or even `delete_all`?